src/realm/alloc_slab.cpp

```diff
@@ -757,32 +757,73 @@ ref_type SlabAlloc::attach_file(const std::string& file_path, Config& cfg)
         size = initial_size;
     }
     ref_type top_ref;
-    File::Map<char> initial_mapping;
+    int file_format_version;
     try {
-        File::Map<char> map(m_file, File::access_ReadOnly, size); // Throws
         note_reader_start(this);
         // we'll read header and (potentially) footer
-        realm::util::encryption_read_barrier(map, 0, sizeof(Header));
-        realm::util::encryption_read_barrier(map, size - sizeof(Header), sizeof(Header));
-
-        validate_header(map.get_addr(), size, path); // Throws
-
-        top_ref = get_top_ref(map.get_addr(), size);
-
-        m_data = map.get_addr();
-        initial_mapping = std::move(map); // replace at end of function
-        // with correctly sized chunks instead...
-        m_baseline = 0;
+        File::Map<char> map_header(m_file, File::access_ReadOnly, sizeof(Header));
+        // if file is too small we'll catch it in validate_header - but we need to prevent an invalid mapping first
+        size_t footer_ref = size < (sizeof(StreamingFooter) + sizeof(Header)) ? 0 : (size - sizeof(StreamingFooter));
+        size_t footer_page_base = footer_ref & ~(page_size() - 1);
+        size_t footer_offset = footer_ref - footer_page_base;
+        File::Map<char> map_footer(m_file, footer_page_base, File::access_ReadOnly,
+                                   sizeof(StreamingFooter) + footer_offset, 0);
+        realm::util::encryption_read_barrier(map_header, 0, sizeof(Header));
+        realm::util::encryption_read_barrier(map_footer, footer_offset, sizeof(StreamingFooter));
+        auto header = reinterpret_cast<const Header*>(map_header.get_addr());
+        auto footer = reinterpret_cast<const StreamingFooter*>(map_footer.get_addr() + footer_offset);
+        top_ref = validate_header(header, footer, size, path); // Throws
         m_attach_mode = cfg.is_shared ? attach_SharedFile : attach_UnsharedFile;
+        m_data = map_header.get_addr(); // <-- needed below
+
+        // Make sure the database is not on streaming format. If we did not do this,
+        // a later commit would have to do it. That would require coordination with
+        // anybody concurrently joining the session, so it seems easier to do it at
+        // session initialization, even if it means writing the database during open.
+        if (cfg.session_initiator && is_file_on_streaming_form(*header)) {
+            // Don't compare file format version fields as they are allowed to differ.
+            // Also don't compare reserved fields.
+            REALM_ASSERT_EX(header->m_flags == 0, header->m_flags, get_file_path_for_assertions());
+            REALM_ASSERT_EX(header->m_mnemonic[0] == uint8_t('T'), header->m_mnemonic[0],
```

Comment:
We should really schedule a project to go through every REALM_ASSERT and see which should really be done in release builds...

(Not asking you to change this today, since this code is exactly as broken as it was before)